src/network.hpp

```diff
@@ -0,0 +1,119 @@
+#pragma once
+
+// opencv
+#include <opencv2/core.hpp>
+
+// spdlog
+#include <spdlog/spdlog.h>
+
+// std
+#include <algorithm>
+#include <random>
+#include <vector>
+
+// tuple holding information about input of a training data and expected output
+// out of it
+struct Sample {
+public:
+  Sample() = default;
+  Sample(const std::vector<double> &input,
+         const std::vector<double> &expected_output)
+      : input(input), expectedOutput(expected_output) {}
+
+public:
+  std::vector<double> input;
+  std::vector<double> expectedOutput;
+};
+
+// Represents simple feed forward Neural Network.
+// As a cost function we use cross entropy and neuron type is sigmoid neurons.
+// It supports learning by SGD and simple L2 regularization technique.
+class Network {
+public:
+  // Initializes network layers, biases, and weights using the given layer
+  // sizes. Weights are initialized with Xavier initialization for training
+  // stability.
+  Network(const std::vector<u_int32_t> &layers_sizes);
+
+  // Returns a vector where the index of the highest value represents the
+  // predicted answer. The `input` size must match the size of the network's
+  // input layer.
+  std::vector<double> FeedForward(const std::vector<double> &input) const;
```

Comment:
If we know the networks input layer size at compile time we could use an array with length specified by a template parameter.

Reply:
I think that in the future this input is going to change to matrix because it is native container of images that we're going to classify by this network. I think optimalization changes should be done later in development when we're sure that is the best way to proccess data.